core/primitives-core/src/types.rs

```diff
@@ -45,3 +45,70 @@ pub type ReceiptIndex = usize;
 pub type PromiseId = Vec<ReceiptIndex>;
 
 pub type ProtocolVersion = u32;
+
+/// The shard identifier. It may be a arbitrary number - it does not need to be
+/// a number in the range 0..NUM_SHARDS. The shard ids do not need to be
+/// sequential or contiguous.
+///
+/// The shard id is wrapped in a newtype to prevent the old pattern of using
+/// indices in range 0..NUM_SHARDS and casting to ShardId. Once the transition
+/// if fully complete it potentially may be simplified to a regular type alias.
+///
+/// TODO get rid of serde
+#[derive(
+    arbitrary::Arbitrary,
+    borsh::BorshSerialize,
+    borsh::BorshDeserialize,
+    serde::Serialize,
+    serde::Deserialize,
+    Hash,
+    Clone,
+    Copy,
+    Debug,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+)]
+pub struct ShardId(u64);
```

Comment:
Can this be easily changed to `ShardId(u32)`?

Reply:
easily - likely not
is it desirable - absolutely
am I doing it in this PR - absolutely not :)

Reply:
the reason it may not be easy is because it may be used either on chain or in the db and changing it would require a protocol upgrade or some custom serialization to keep it in u64 format when serialized